tests/warp/warp_test.go

```diff
@@ -375,4 +366,106 @@ var _ = ginkgo.Describe("[Warp]", ginkgo.Ordered, func() {
+		os.Setenv("SOURCE_CHAIN_ID", blockchainIDA.Hex())
+		os.Setenv("PAYLOAD", common.Bytes2Hex(payload))
+		os.Setenv("EXPECTED_UNSIGNED_MESSAGE", hex.EncodeToString(expectedUnsignedMessage.Bytes()))
+		runWarpHardhatTests(ctx, rpcURI, cmdPath, testPath)
+
+		log.Info("Waiting for new block confirmation")
+		newHead := <-newHeads
+		blockHash := newHead.Hash()
+
+		log.Info("Fetching relevant warp logs from the newly produced block")
+		logs, err := chainAWSClient.FilterLogs(ctx, interfaces.FilterQuery{
+			BlockHash: &blockHash,
+			Addresses: []common.Address{warp.Module.Address},
+		})
+		gomega.Expect(err).Should(gomega.BeNil())
+		gomega.Expect(len(logs)).Should(gomega.Equal(1))
+
+		// Check for relevant warp log from subscription and ensure that it matches
+		// the log extracted from the last block.
+		txLog := logs[0]
+		log.Info("Parsing logData as unsigned warp message")
+		unsignedMsg, err := warp.UnpackSendWarpEventDataToMessage(txLog.Data)
+		gomega.Expect(err).Should(gomega.BeNil())
+
+		// Set local variables for the duration of the test
+		unsignedWarpMessageID = unsignedMsg.ID()
+		unsignedWarpMsg = unsignedMsg
+		log.Info("Parsed unsignedWarpMsg", "unsignedWarpMessageID", unsignedWarpMessageID, "unsignedWarpMessage", unsignedWarpMsg)
+
+		// Loop over each client on chain A to ensure they all have time to accept the block.
+		// Note: if we did not confirm this here, the next stage could be racy since it assumes every node
+		// has accepted the block.
+		for i, uri := range chainAURIs {
+			chainAWSURI := toWebsocketURI(uri, blockchainIDA.String())
+			log.Info("Creating ethclient for blockchainA", "wsURI", chainAWSURI)
+			client, err := ethclient.Dial(chainAWSURI)
+			gomega.Expect(err).Should(gomega.BeNil())
+
+			// Loop until each node has advanced to >= the height of the block that emitted the warp log
+			for {
+				block, err := client.BlockByNumber(ctx, nil)
+				gomega.Expect(err).Should(gomega.BeNil())
+				if block.NumberU64() >= newHead.Number.Uint64() {
+					log.Info("client accepted the block containing SendWarpMessage", "client", i, "height", block.NumberU64())
+					break
+				}
+			}
+		}
+	})
 })
+
+func runWarpHardhatTests(ctx context.Context, chainURI string, execPath string, testPath string) {
+	log.Info(
+		"Executing HardHat tests on warp blockchain",
+		"testPath", testPath,
+		"ChainURI", chainURI,
+	)
+
+	cmd := exec.Command("npx", "hardhat", "test", testPath, "--network", "local")
+	cmd.Dir = execPath
+
+	log.Info("Sleeping to wait for test ping", "rpcURI", chainURI)
+	err := os.Setenv("RPC_URI", chainURI)
+	gomega.Expect(err).Should(gomega.BeNil())
+	log.Info("Running test command", "cmd", cmd.String())
+
+	out, err := cmd.CombinedOutput()
+	fmt.Printf("\nCombined output:\n\n%s\n", string(out))
+	gomega.Expect(err).Should(gomega.BeNil())
+}
+
+func toRPCURI(uri string, blockchainID string) string {
+	return fmt.Sprintf("%s/ext/bc/%s/rpc", uri, blockchainID)
```

Comment:
Can we use `utils.RunHardHatTests` here instead? It looks like the only difference between the two is logs, which we can move outside of this function if it's important to include extra logs that specify this is warp related.

Reply:
`RunHardHatTests` uses a default port (9650) whereas we use a custom one in these tests. Added a new helper method and reduced duplicate code.